src/SSO.php

```diff
@@ -37,84 +47,46 @@ public function __construct(Request $request, User $user, Config $config)
         $this->request = $request;
         $this->config = $config;
         $this->user = $user;
-    }
-    /**
-     * Checks if the request client id is equal to the config client id
-     *
-     * @return boolean
-     */
-    protected function isInvalidClientID()
-    {
-        return $this->request->getClientID() !== $this->config->getClientID();
-    }
 
-    /**
-     * Checks that the request timestamp and signature are set
-     *
-     * @return boolean
-     */
-    protected function isUnsigned()
-    {
-        $requestTimestamp = $this->request->getTimestamp();
-        $requestSignature = $this->request->getSignature();
+        $this->addValidator(new Validator\MissingClientID);
+        $this->addValidator(new Validator\InvalidClientID);
+        $this->addValidator(new Validator\UnsignedRequest);
+        $this->addValidator(new Validator\InvalidTimestamp);
+        $this->addValidator(new Validator\MissingSignature);
+        $this->addValidator(new Validator\ExpiredTimestamp);
+        $this->addValidator(new Validator\InvalidSignature);
 
-        return !isset($requestTimestamp) && !isset($requestSignature);
     }
 
     /**
-     * Signature is valid if the request and new hash are equal
+     * Adds custom external validation functions to run in addition to the Vanilla core
      *
-     * @return boolean
+     * @param function
      */
-    protected function isSignatureValid()
+    public function addValidator($validator)
     {
-        $timestamp = $this->request->getTimestamp();
-        $secret = $this->config->getSecret();
-        $signature = md5($timestamp.$secret);
-        return $this->request->getSignature() === $signature;
+        if (is_callable($validator)) {
+            $this->validators[] = new Validator\Closure($validator);
+        } elseif ($validator instanceof ValidatorInterface) {
+            $this->validators[] = $validator;
+        }
```

Comment:
@j-feng13 We should throw `\InvalidArgumentException` if the user passes in something that's not callable or not a `ValidatorInterface`.  Also, you should check `is_object($validator) && $validator instanceof ValidatorInterface` to prevent PHP from complaining that `$validator` is not an object (in the case that the user passed in something weird.)

You can do this in another PR.